src/include/miopen/nllloss/problem_description.hpp

```diff
@@ -0,0 +1,261 @@
+        {
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
+            MIOPEN_THROW(miopenStatusBadParm, "NLLLoss: Unpacked tensors not supported.");
+#else
+            return false;
+#endif
+        }
+        return true;
+    }
+
+protected:
+    TensorDescriptor inputDesc;
+    TensorDescriptor targetDesc;
+    TensorDescriptor weightDesc;
+    TensorDescriptor outputDesc;
+
+    int32_t ignore_index;
+    bool is_fwd;
+
+    NetworkConfig MakeForwardNetworkConfig() const;
+};
+
+struct UnreduceProblemDescription : ProblemDescription
+{
+    UnreduceProblemDescription(const TensorDescriptor& inputDesc_,
+                               const TensorDescriptor& targetDesc_,
+                               const TensorDescriptor& weightDesc_,
+                               const TensorDescriptor& outputDesc_,
+                               int32_t ignore_index_,
+                               bool is_fwd_)
+        : ProblemDescription(
+              inputDesc_, targetDesc_, weightDesc_, outputDesc_, ignore_index_, is_fwd_)
+    {
+        IsSameType();
+        IsRightLength();
+        IsRightStride();
+        IsAllPacked();
+    }
+
+    size_t GetNtotal() const { return outputDesc.GetElementSize(); }
+    size_t GetC() const { return weightDesc.GetElementSize(); }
+
+    bool IsSameType() const
+    {
+        if(inputDesc.GetType() != weightDesc.GetType())
+        {
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "NLLLoss: Tensor types of Input and Weight do not match.");
+#else
+            return false;
+#endif
+        }
+        return true;
+    }
+
+    NetworkConfig MakeNetworkConfig() const override;
+
+private:
+    NetworkConfig MakeForwardNetworkConfig() const;
+};
+
+struct ReduceProblemDescription : ProblemDescription
+{
+    ReduceProblemDescription(const TensorDescriptor& inputDesc_,
+                             const TensorDescriptor& targetDesc_,
+                             const TensorDescriptor& weightDesc_,
+                             const TensorDescriptor& outputDesc_,
+                             int32_t ignore_index_,
+                             bool is_fwd_)
+        : ProblemDescription(
+              inputDesc_, targetDesc_, weightDesc_, outputDesc_, ignore_index_, is_fwd_)
+    {
+        IsSameType();
+        IsRightLength();
+        IsRightStride();
+        IsAllPacked();
+    }
+
+    size_t GetNtotal() const { return targetDesc.GetElementSize(); }
+    size_t GetC() const { return weightDesc.GetElementSize(); }
+
+    bool IsSameType() const
+    {
+        if(inputDesc.GetType() != weightDesc.GetType())
+        {
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "NLLLoss: Tensor types of Input and Weight do not match.");
+#else
+            return false;
+#endif
+        }
+        return true;
+    }
+
+    bool IsRightLength() const
```

Comment:
change right to valid and remove return refer https://github.com/ROCm/MIOpen/pull/2766#discussion_r1575181661